### ISSUE TITLE
Add new GitHub Actions workflow caring out the cypress test suite

### ISSUE THREAD

**Issue by Ein-Tim** [corona-warn-app/cwa-website#2892]
## Feature description

A new GitHub Actions workflow caring out the cypress test suite should be added to this repository. The workflow should run for all opened PRs in this repository, like the "build and test" workflow does now. 

## Problem and motivation

- PR https://github.com/corona-warn-app/cwa-website/pull/2891 added the (strong) recommendation to carry out the cypress test suite before opening PRs in this repository.
It would be easier for everyone, the contributors as well as the Open-Source-Team if the tests are carried out automatically, without having to start them manually.

**Comment by MikeMcC399** [comment]
@Ein-Tim 

It would be problematic to include the `check_links.js` Cypress test in a workflow, since this test depends on links to external websites and their state. `cwa-website` has no control over the external websites, so this could cause unpredictable failures.

@dsarkar's https://github.com/corona-warn-app/cwa-website/issues/1825#issuecomment-935668508 about the run time of the test being 10 minutes is also still relevant.

**Comment by Ein-Tim** [comment]
@MikeMcC399 

> It would be problematic to include the `check_links.js` Cypress test in a workflow, since this test depends on links to external websites and their state. `cwa-website` has no control over the external websites, so this could cause unpredictable failures.

I don't think that this is a problem. The check should not be required to pass to be able to merge the PR, but rather feedback for the PR author. If the check fails, the PR author can check if this is an issue on our side or if this is an issue of an external website (which can then be fixed in another PR, by excluding the website). 

> @dsarkar's [#1825 (comment)](https://github.com/corona-warn-app/cwa-website/issues/1825#issuecomment-935668508) about the run time of the test being 10 minutes is also still relevant.

This argument seems outdated. If everyone follows the recommendation to run the Cypress test suite locally, it will take 10 minutes on their device. If it runs on GitHub, it takes 10 minutes on GitHub. There are no time savings either way.

**Comment by MikeMcC399** [comment]
@Ein-Tim 

This suggestion could be implemented step-by-step. 

Some ideas:

- Add all Cypress tests, except [check_links.js](https://github.com/corona-warn-app/cwa-website/blob/master/cypress/integration/check_links.js) into the [build](https://github.com/corona-warn-app/cwa-website/blob/master/.github/workflows/build.yml) workflow.
- Also make the Cypress tests run periodically in a different workflow.
- Split the checks in [check_links.js](https://github.com/corona-warn-app/cwa-website/blob/master/cypress/integration/check_links.js) into:
    - check local relative links
    - check external absolute links
- then add check local relative links to the workflow

The current manual running of Cypress tests is more flexible, however it requires discipline on the part of PR submitters. The flexibility comes for instance in being able to choose whether or not to re-run the whole test suite if a minor change is made to a PR.

I will see if I can make some more concrete proposals. If this is realized then it would probably need to be a team effort.

Typo: "caring out" in the title is probably meant to be "carrying out" as in "to execute"

**Comment by Ein-Tim** [comment]
FYI: Here is the Cypress documentation on implementing cypress tests into a GitHub Actions workflow: https://docs.cypress.io/guides/continuous-integration/github-actions?utm_source=Test%20Runner&utm_medium=CI%20Prompt%201&utm_campaign=GitHub&utm_content=Automatic#What-you-ll-learn

I tried to add the example workflow they presented there to my repository (https://github.com/Ein-Tim/cwa-website/tree/test/cypress-GH-action), however, the tests always failed as Cypress was not able to verify that the localhost server is actually up and running (https://github.com/Ein-Tim/cwa-website/actions/runs/2410739771 & https://github.com/Ein-Tim/cwa-website/actions/runs/2410756914).

So it seems like this would need only minor effort to implement for someone with GitHub Actions development skills. The only problem is that the test takes very long, but as described [here](https://www.youtube.com/watch?v=96Yn_IiQUJI), it is possible to run different tests in parallel.

@svengabr Would you like to also bring this up to the next team meeting? Especially if tests are run in parallel, and only the most important tests are run (checking links, etc.), I think this is would be a great improvement!

**Comment by dsarkar** [comment]
@Ein-Tim @MikeMcC399 Thanks for suggestions and discussion this here. We will have a look!

**Comment by svengabr** [comment]
@Ein-Tim @MikeMcC399 our suggestion is also to run the test on a scheduled basis for example:
```
on:
  schedule:
    - cron: '0 10 * * *'
    - cron: '0 16 * * *'
```
This will trigger the tests at 10:00 and 16:00 each day.

We can also include some tests on every PR but not all tests at the same since running all tests at once will take too much time. I'm not even sure if it makes sense to have tests on every PR. The regular build process will make sure the code is ok and we should be okay with the scheduled runs twice a day.

Especially on busy days like release day, we need to merge multiple PRs in a short period of time where we cant wait ages each time. Even if a test was successful previously - an `update from main` will re-trigger everything and we would have to wait again.

**Comment by Ein-Tim** [comment]
@svengabr

Thanks for your feedback! Scheduled tests seem fine to me, however, I wonder how long the tests would take if run in parallel. Probably still too long because of the link checking test...

I would welcome running the check on a scheduled basis, if this means that I don't have to run the test suite every time locally on my device when providing a PR.

**Comment by Ein-Tim** [comment]
@MikeMcC399 

> I noticed you mentioned in another issue that it took one hour in your environment. That does not sound normal.

Yesterday, on PR #2921, running the Cypress tests (`npm run test:open`) took ca. 30 minutes, but as my machine was under heavy workload at this time this is expected, or?
I'll give it another try today, without my machine having to perform any other tasks.


> It would still be good practice to run `npm test` locally as part of a PR submission to avoid merging something which could be broken. A scheduled test run from the master branch would be a second line of defence.

Agreed. But just for clarification, what should I do when providing a PR? Running `npm test` or running all tests with `npm run test:open`?

**Comment by MikeMcC399** [comment]
@Ein-Tim 

> what should I do when providing a PR? Running `npm test` or running all tests with `npm run test:open`?

Perhaps the description in https://github.com/corona-warn-app/cwa-website#testing is not so clear?

`npm test` builds the web and runs all tests automatically using the Chrome browser.

`npm run test:open` also builds the web, then it opens up the Cypress test runner. 

The Cypress test runner gives you more flexibility compared to `npm test`: you can choose which browser to use, choose which test to run, or if you click on "Run n integration tests", all tests will be run.

So unless you have special needs, you can use `npm test` (or just type `npm t`).

**Comment by Ein-Tim** [comment]
@MikeMcC399 Ah, okay, I understand! So in the end, both `npm test` & `npm run test:open` do the same?

**Comment by MikeMcC399** [comment]
@Ein-Tim 

> So in the end, both npm test & npm run test:open do the same?

`npm test` always runs all tests. 
`npm run test:open` can be used to run all tests

They are not the same, but can both be used to run all tests.

**Comment by Ein-Tim** [comment]
@MikeMcC399 Okay, thanks!

**Comment by Ein-Tim** [comment]
So, I just made the test. Running `npm run test:open` and then selecting to carry out all tests takes ca. 25 minutes on my machine. The bottleneck here is that Cypress doesn't run natively on my M1 MacBook but has to be emulated via Rosetta 2. Thus also the browser (Microsoft Edge), which normally runs natively on M1 chips, is emulated via Rosetta 2.

**Comment by svengabr** [comment]
Ok, how do we move forward with this? Will you provide a PR adding the new `.github/workflow` or should someone from my team add it?

**Comment by Ein-Tim** [comment]
@svengabr

It would be great if someone from the team could add it! Because currently, just adding the workflow provided by Cypress doesn't seem to be successful.

**Comment by MikeMcC399** [comment]
@svengabr 

I have submitted:

- https://github.com/corona-warn-app/cwa-website/pull/2930 and
- https://github.com/corona-warn-app/cwa-website/pull/2932

to support implementing the workflow issue.

I'm also quite close to having a simple workflow implemented to do the tests. I will share with you when I have it tested.

**Comment by svengabr** [comment]
Thanks, @MikeMcC399 !

The team was a bit busy the last few days because of the 2.23 release. I have edited your PRs to merge into the branch `cypress-workflow` so we can review it as a whole while also adding the new Github workflow.

**Comment by MikeMcC399** [comment]
@svengabr 

> The team was a bit busy the last few days because of the 2.23 release. 

Sure. I understand that, which is why I waited until today before submitting these two PRs.

> I have edited your PRs to merge into the branch `cypress-workflow` so we can review it as a whole while also adding the new Github workflow.

Great decision!

I have a new Cypress test prepared which tests the links used by the Android / iOS apps. The workflow is not dependent on this test, so I could submit a PR to master, but maybe you would prefer a PR into the `cypress-workflow` branch for easier reviewing? Could you please let me know which branch you prefer?

**Comment by MikeMcC399** [comment]
@svengabr 

> It would be easier to have it merged to cypress-workflow first so we can review it as a whole.

- Thanks for the advice. I have now submitted PR https://github.com/corona-warn-app/cwa-website/pull/2935 for your consideration. It fills a gap where the links used by the apps were not specifically tested.

**Comment by MikeMcC399** [comment]
- I submitted PR #2936 as a draft suggestion to implement a scheduled workflow testing the production server and running tests except for `check_links.js`.
